### PR TITLE
fix: [CDS-49140]: CommandExecutionStatus will not change once it has reached terminal status.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
By default, when the log stream is closed and we try to write into that steam the command execution status switches back to running state. To avoid this I have added a check.

## Automation
UT Added

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java [hunk 1]
```diff
@@ -91,6 +91,9 @@ boolean updateCommandUnitProgressMap(CommandExecutionStatus commandExecutionStat
       change = true;
     } else {
       CommandUnitProgress commandUnitProgress = commandUnitProgressMap.get(commandUnitName);
+      if (CommandExecutionStatus.isTerminalStatus(commandUnitProgress.getStatus())) {
```
Comment (vaibhav-si) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java:94]: Add a warning here to indicate to the caller they are doing something wrong.

Once the status has been marked success for a unit, they should not send logs again on that unit.

Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java:94]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java [hunk 2]
```diff
@@ -91,6 +91,10 @@ boolean updateCommandUnitProgressMap(CommandExecutionStatus commandExecutionStat
...
+        log.warn("Unable to write to the log stream as it is already closed for command unit: {}", commandUnitName);
```
Comment (vaibhav-si) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java:95]: Log should be instead "Skipping updating command unit status as the unit has already reached terminal status"

Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java]: Done

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)* status: 2 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)* status: 3 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by GoelPratyush99** [comment]
TI-bootstrap1: trigger ti1

**Comment by GoelPratyush99** [comment]
trigger ti4

**Comment by GoelPratyush99** [comment]
trigger ti1

**Review by abosii-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)* status: 2. unresolved discussions.

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)* status: 2. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java` line 94 at r1](https://reviewable.io/reviews/harness/harness-core/41937#-NJxX1Wn5tdSKBwf1LyW-r1-94:-NK-H1fR5jf9eHPgyOvE:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/cf3bb9f35680d5c2bf2e31055ae3619b93328d4f/930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java#L94)):*
<details><summary><i>Previously, GoelPratyush99 (Pratyush Goel) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java` line 95 at r2](https://reviewable.io/reviews/harness/harness-core/41937#-NJxX1Wn5tdSKBwf1LyW-r2-95:-NK-H2YzCVGcQMDsS0jN:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/9f8e2d2ba751080199e890acb2175d3583d5b2cc/930-delegate-tasks/src/main/java/io/harness/delegate/beans/logstreaming/NGDelegateLogCallback.java#L95)):*
<details><summary><i>Previously, GoelPratyush99 (Pratyush Goel) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Done

**Review by vaibhav-si — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41937)* status: :shipit: complete! ✓

**Pull request merged**